### PR TITLE
Add auth proc filters hook

### PR DESCRIPTION
**Pull request by gmrsagar**
Addressing the issue #388

### CODE REVIEW

File: config/config.php [hunk 1]
```diff
@@ -88,12 +88,7 @@
 
     'proxy' => null, // TODO inherit from moodle conf see http://moodle.local/admin/settings.php?section=http for more.
 
-    'authproc.sp' => array(
-        50 => array(
-            'class' => 'core:AttributeMap',
-            'oid2name',
-        ),
-    ),
+    'authproc.sp' => auth_plugin_saml2::saml2_authproc_filters_hook(),
```
Comment (brendanheywood) [inline config/config.php]: look good, but please don't remove the core:AttributeMap one that was already there. Or move this into saml2_authproc_filters_hook

File: auth.php [hunk 2]
```diff
@@ -104,6 +104,19 @@ public function __construct() {
         $this->defaultidp = auth_saml2_get_default_idp();
     }
 
+    public static function saml2_authproc_filters_hook()
+    {
+        if (function_exists('get_plugins_with_function')) {
```
Comment (brendanheywood) [inline auth.php]: is this if needed?

File: auth.php [hunk 3]
```diff
@@ -104,6 +104,19 @@ public function __construct() {
...
+            $callbacks = get_plugins_with_function('extend_auth_saml2_proc', 'lib.php');
+            foreach ($callbacks as $type => $plugins) {
+                foreach ($plugins as $plugin => $pluginfunction) {
+                    $authprocfilters = $pluginfunction();
```
Comment (brendanheywood) [inline auth.php]: I think you want to keep appending more items onto the array, and also start the array with an explicit empty array

File: config/config.php (hunk 1 above)
Comment (gmrsagar) [reply config/config.php]: moved this to saml2_authproc_filters_hook as a default hook

File: auth.php (hunk 2 above)
Comment (gmrsagar) [reply auth.php]: you're right this is not needed. removed it.

File: auth.php (hunk 3 above)
Comment (gmrsagar) [reply auth.php]: updated with a default filter.

File: auth.php [hunk 4]
```diff
@@ -104,6 +104,37 @@ public function __construct() {
...
+            'class' => 'core:AttributeMap',
+            'oid2name',
+        );
+        $callbacks = get_plugins_with_function('extend_auth_saml2_proc', 'lib.php');
+        foreach ($callbacks as $plugins) {
+            foreach ($plugins as $pluginfunction) {
+                $filters = $pluginfunction();
+                foreach ($filters as $key => $value) {
+                    $key = self::check_filters_priority($key, $authprocfilters);
+                    $authprocfilters[$key] = $value;
```
Comment (brendanheywood) [inline auth.php]: can you please emit a debugger warning when two proc filters have the same key and clash

File: README.md [hunk 5]
```diff
@@ -276,6 +276,33 @@ Suggested attribute mappings:
 |`Email`|`user.email`|
 
 
+**Auth Proc Filter Hooks**
+
+Other plugins may hook into SAML2 and create custom Auth Proc Filters.
+Auth Proc Filters allows you to do stuff at the IdP after authentication is complete
```
Comment (brendanheywood) [inline README.md]: This isn't quite correct

"Auth Proc Filters allows you to do stuff at the IdP after authentication is complete"

something like

"Auth Proc Filters allows you to mutate of the attributes passed back from the IdP before Moodle handles them and does the profile field mappings"

File: README.md [hunk 6]
```diff
@@ -276,6 +276,33 @@ Suggested attribute mappings:
...
+
+Steps to implement the hook:
+* Create a plugin that will implement the hook (e.g local_hookimplement)
+* Define the hook function 'local_hookimplement_extend_auth_saml2_proc' in plugin's lib.php
+* The function should return array of SimpleSaml Auth Proc Filters.
+
+Example:
+```php
+function local_hookimplement_extend_auth_saml2_proc {
+   // Do your data validation
```
Comment (brendanheywood) [inline README.md]: I'd remove this line

File: README.md [hunk 7]
```diff
@@ -276,6 +276,33 @@ Suggested attribute mappings:
...
+
+   return [
+      51 => array(
+         'class' => 'core:AttributeMap',
+         'oid2name'
+      )
+   ]
```
Comment (brendanheywood) [inline README.md]: can you please add a more realistic example with the core:php

**Pull request merged**